### DISCUSSION TITLE
DIY Geiger counter and Tasmota ESP8266

### DISCUSSION THREAD

**Discussion by Kantulak** [arendst/Tasmota#18449] [Show and tell]
Hi,

I want to share a solution for DIY Geiger counter integration using Tasmota running on ESP8266.
Steps:
1. Configure the connected GPIO as Counter_n (1) 
2. Set TelePeriod 60 
3. Set SetOption79 1  - to reset the counter every 60 seconds
4. Add and enable the script:
>D
c=0
s=0
>T
c=pc[1]
;J305ß Geiger Mueller tube background noise: 12 Counts per Minute
;J305ß factor: 0.00812037037037
s=c-12
if s>=0 { 
s=s*0.008120
} else {
s=0
}

>J
,"GMC":{"Radiation":%s%}
#

5. With the proper MQTT settings and thanks to integration of Tasmota in Home Assistant, 
after some time you will find 2 new sensors on your device: 
"COUNTER C1" - raw Counts per Minute (CPM)
"GMC Radiation" in uSv/hr

I hope someone improves this simple solution.

**Comment by vic42** [comment]
Cool, I chose a different path for adapting a MightyOhm Geiger Counter: The **serial bridge** configuration.
![Screenshot_20231106_182455](https://github.com/arendst/Tasmota/assets/3363362/f1504ac6-0296-4071-801d-bf215ba1630a)

It gives me these MQTT messages that I dissect later in node-red:
`tele/geiger/RESULT {"SSerialReceived":"CPS, 0, CPM, 20, uSv/hr, 0.11, SLOW\r\n"}`

Function node code:
```
var str = msg.payload.SSerialReceived;
var array = str.split(",");
var tarray = array.map(string => string.trim());
var msv = tarray[5];
msg.topic = "radiation";
msg.payload = parseFloat(msv);
return msg;
```
![Screenshot_20231106_183351](https://github.com/arendst/Tasmota/assets/3363362/a4b070bb-9454-4f94-9ab8-5175ab5e1c6d)

**Comment by raindancer2204** [comment]
Hi, I would also be interested in the solution with the DIY Geiger counter. The MightyOhm Geiger is too expensive for me. Are there other DIYs with a serial port for Tastmota? Can you please tell me which DIY you used. Thank you very much

**Reply by vic42** [reply]
I simply connected power and serial of the MightyOhm kit to a Wemos D1 mini, and that was it. I guess pretty much any Geiger counter will do, be it pulse or serial based. I'd consult my preferred search engine.

https://www.instructables.com/Simplest-Geiger-Counter/
http://www.galacticelectronics.com/GeigerCounter.HTML
https://sites.google.com/site/diygeigercounter//

It basically consists of a low current 400V~500V power supply, a GM tube and and a transistor or OpAmp to detect the events on the tube's cathode.

Amazon has kits for under $40, including the tube.

**Reply by cleverpal-ai** [reply]
How did you configure Tasmota / WeMos D1 mini for this DIY Geiger counter? 
Do I simply need to connect it to ESP32 GPIO  and 5V/GND or I need some extra manipulations to get it working?

**Reply by Kantulak** [reply]
Hi, for DIY just search for "Assembled DIY Geiger Counter Kit Module" or similar - priced at 30-35. No serial connection, Tasmota counts pulses (one wire). All you need is to power up your module and connect 1 GPIO to counter. I am using cheap step down converter 5V->3.3V to power ESP-01s module right from the counter. So, you need 1. kit, 2. cheap tasmotized module, 3. stepdown converter (optional) and some wires.  And it works pretty reliable.

**Reply by Kantulak** [reply]
step down: AMS1117
![GMC_ESP01](https://github.com/user-attachments/assets/4e2d2f14-369a-4035-afa9-1a6cb68cffaf)

**Reply by cleverpal-ai** [reply]
@Kantulak Thanks! I have connected 1 GPIO to Counter and it does count, but I can't run the Script you posted. Where do I need to save/load it? I tried via Berry Script console, but it says:
_BRY: Exception> 'syntax_error' - input:1: unexpected symbol near '>'_

```
>D
  c=0
  s=0

>T
  c=pc[1]
  ; J305ß Geiger Mueller tube background noise: 12 Counts per Minute
  ; J305ß factor: 0.00812037037037
  s=c-12
  if s>=0 {
    s=s*0.00812037037037
  } else {
    s=0
  }

>J
  {"GMC":{"Radiation":%s%}}
```

**Reply by vic42** [reply]
`BRY:` means you are probably trying to run it in berry on an ESP32 MCU. This script is made for an older scripting language, typically used with ESP8266 chips. So either you learn how to convert it to berry or you need to build a binary including the old scripting language.
See
* https://tasmota.github.io/docs/Scripting-Language/
* https://tasmota.github.io/docs/Berry/
* https://tasmota.github.io/docs/Compile-your-build/

for details.

**Reply by sfromis** [reply]
Also beware that it looks like you may be using a 5V signal to the ESP8266, which is not within specs. I'm not saying "instant death", more like a risk factor, and maybe shorter lifetime.

**Reply by sfromis** [reply]
Here's a bit of Berry code, aiming to do the same as that scripter code:
```
class GeigerDriver
  var radiation
  def init()
    tasmota.remove_driver(global.GeigerDriver_instance)
    tasmota.remove_rule('Counter#C1')
    tasmota.add_rule('Counter#C1',
      def (value)
        self.radiation = (value == 0) ? nil : (value - 12) * 0.00812037037037
      end)
    # J305ß Geiger Mueller tube background noise: 12 Counts per Minute
    # J305ß factor: 0.00812037037037
    tasmota.add_driver(global.GeigerDriver_instance := self)
  end
  def web_sensor()
    if self.radiation
      tasmota.web_send_decimal(format('{s}GMC Radiation{m}%i{e}', self.radiation))
    end
  end
  def json_append()
    if self.radiation
      tasmota.response_append(f',"GMC":{{"Radiation":{self.radiation}}}')
    end
  end
end
gd = GeigerDriver()
```
No custom build needed.

Whether the calculation is correct, I have no input on.

**Reply by 420pootang69** [reply]
> Also beware that it looks like you may be using a 5V signal to the ESP8266, which is not within specs. I'm not saying "instant death", more like a risk factor, and maybe shorter lifetime.

Just looking at the specs for my board, that 5V is coming through over 500K worth of resistors. Would the current be so low (~0.01mA) as it wouldn't be an issue?

I remember it was a thing with the PIC16 devices that you could connect them direct to mains with a large resistor (for measuring mains frequency) , and although the voltage was obviously way over spec, it was fine as it's mainly current that would damage them.

**Comment by lalo-uy** [comment]
Pic16 gpio have protection diodes to vcc & gnd capable of 2mA.
Thats why they can be connected to mains thru 470k and stand less than 1mA.

Esp do not have that kind of protection.

**Comment by 420pootang69** [comment]
Thank you!
